src/nccl_ofi_rdma.c

```diff
@@ -2622,6 +2636,70 @@ static ncclResult_t reg_mr_ep(nccl_net_ofi_rdma_ep_t *ep, void *data,
 	return ret;
 }
 
+/*
+ * @brief	Register memory region on RDMA endpoint
+ *
+ * When a process executes the fork() syscall, all process memory pages
+ * are marked as CoW (copy-on-write) such that the virtual pages are
+ * read-only on both parent and child processes and when one of them
+ * writes to a page, a page-fault is triggered which cause OS to copy the
+ * page to a new physical page and change virtual page to be mapped to
+ * the new physical page with writable access.
+ *
+ * In order for MRs to properly be used as device DMA source/target,
+ * their physical pages must be pinned. In order to avoid changing MRs
+ * physical pages after a fork(), rdma-core historically
+ * madvice(MADV_DONTFORK) their buffers. fork() handles memory pages
+ * marked with MADV_DONTFORK by keeping them writable on parent and
+ * providing new zeroed physical pages on child.
+ *
+ * This assumes that the content of a page marked with MADV_DONTFORK is
+ * not used by the child. However, this assumption is wrong when a MR do
+ * not cover the entire page, because the remainder of the page may
+ * contain content that the child intends to use. Which may lead to
+ * various hard to debug issues in the child process (e.g. memory
+ * corruption on CRT heap).
+ *
+ * To address this issue, kernel 5.15 introduced copy-on-fork support to
+ * not require userspace to mark any memory page MADV_DONTFORK but
+ * instead kernel copy the content of pinned memory pages from parent to
+ * child immediately when fork() is executed.
+ *
+ * In attempt to avoid this issue in old kernels without copy-on-fork,
+ * we enlarge our MRs to cover full memory pages and assert that this
+ * is the case to avoid introducing such hard to debug issues in the
+ * future. Note that we can only do this though on internal MRs and
+ * NCCL is still allowed to register MRs which do not cover full
+ * memory pages.
+ *
+ * It's worth emphasizing that registering a MR which does not cover a
+ * full memory page on a kernel without copy-on-fork won't necessarily
+ * result in an issue. Because fork() may never be executed, or an
+ * execve() may immediately be executed after fork() such that the above
+ * mentioned issue is not encountered.
+ *
+ * @param	ep
+ *		RDMA endpoint on which memory region is registered
+ * @param	data
+ *		Pointer to MR. MR must be aligned to system memory page size.
+ * @param	size
+ *		Size of MR. Size must be a multiple of system memory page size.
+ * @param	type
+ *		Type of MR
+ *
+ * @return	Memory registration handle
+*/
+static ncclResult_t reg_internal_mr_ep(nccl_net_ofi_rdma_ep_t *ep, void *data,
```

Comment:
It would've been preferred to pass a boolean parameter to `reg_mr_ep()` indicating whether its an internal or external MR buffer, so that it will enforce these asserts only for internal MR buffers. That would force the user of `reg_mr_ep()` to think whether its internal or external, without the possibility of mistakenly call directly `reg_mr_ep()` on an internal MR buffer.